include/simdjson/dom/serialization-inl.h

```diff
@@ -0,0 +1,361 @@
+  write_pointer -= 1;
+  while (second_write_pointer < write_pointer) {
+    char c1 = *write_pointer;
+    char c2 = *second_write_pointer;
+    *second_write_pointer = c1;
+    *write_pointer = c2;
+    write_pointer--;
+    second_write_pointer++;
+  }
+  return answer;
+}
+} // anonymous namespace
+
+
+/***
+ * Minifier/formatter code.
+ **/
+
+simdjson_really_inline void mini_formatter::number(uint64_t x) {
+  char number_buffer[24];
+  char *newp = fast_itoa(number_buffer, x);
+  buffer.insert(buffer.end(), number_buffer, newp);
+}
+
+simdjson_really_inline void mini_formatter::number(int64_t x) {
+  char number_buffer[24];
+  char *newp = fast_itoa(number_buffer, x);
+  buffer.insert(buffer.end(), number_buffer, newp);
+}
+
+simdjson_really_inline void mini_formatter::number(double x) {
+  char number_buffer[24];
+  char *newp = to_chars(number_buffer, nullptr, x);
+  buffer.insert(buffer.end(), number_buffer, newp);
+}
+
+simdjson_really_inline void mini_formatter::start_array() { one_char('['); }
+simdjson_really_inline void mini_formatter::end_array() { one_char(']'); }
+simdjson_really_inline void mini_formatter::start_object() { one_char('{'); }
+simdjson_really_inline void mini_formatter::end_object() { one_char('}'); }
+simdjson_really_inline void mini_formatter::comma() { one_char(','); }
+simdjson_really_inline void mini_formatter::c_str(const char *c) {
+  for (; *c != '\0'; c++) {
+    one_char(*c);
+  }
+}
+
+simdjson_really_inline void mini_formatter::true_atom() { 
+  const char * s = "true";
+  buffer.insert(buffer.end(), s, s + 4);
+}
+simdjson_really_inline void mini_formatter::false_atom() {
+  const char * s = "false";
+  buffer.insert(buffer.end(), s, s + 5);
+}
+simdjson_really_inline void mini_formatter::null_atom() {
+  const char * s = "null";
+  buffer.insert(buffer.end(), s, s + 4);
+}
+simdjson_really_inline void mini_formatter::one_char(char c) { buffer.push_back(c); }
+simdjson_really_inline void mini_formatter::key(std::string_view unescaped) {
+  string(unescaped);
+  one_char(':');
+}
+simdjson_really_inline void mini_formatter::string(std::string_view unescaped) {
+  one_char('\"');
+  size_t i = 0;
+  // fast path for the case where we have no control character, no ", and no backslash (fast path)
+  for (; (i < unescaped.length()) && (uint8_t(unescaped[i]) > 0x1F) 
+        && (unescaped[i] != '\"') && (unescaped[i] != '\\'); i++) {}
+  buffer.insert(buffer.end(), unescaped.data(), unescaped.data() + i);
+  // We caught a control character if we enter this loop (slow)
+  for (; i < unescaped.length(); i++) {
+    switch (unescaped[i]) {
+    case '\"':
+      {
+        const char * s = "\\\"";
+        buffer.insert(buffer.end(), s, s + 2);
+      }
+      break;
+    case '\\':
+      {
+        const char * s = "\\\\";
+        buffer.insert(buffer.end(), s, s + 2);
+      }
+      break;
+    default:
+      if (uint8_t(unescaped[i]) <= 0x1F) {
+        const static char *escaped[] = {
```

Comment:
I'm a little surprised we don't already have this table!

Reply:
I don't think it matters, but if we *can* use `constexpr` it probably communicates intent better. I think.

Since this is a function, how does initialization of this thing work? Is it statically embedded in the data section, or does it end up initializing it the first time you call the function? "It doesn't matter, it'll be just as fast and I kinda don't want to spend time looking this up" is an acceptable answer :)

Reply:
> I'm a little surprised we don't already have this table!

I don't think we do.

I think I will make it a bit faster. Having variable length strings introduces indirection.

>  Is it statically embedded in the data section, or does it end up initializing it the first time you call the function?

Statically embedded.

I don't think that the keyword "static" even matters.